core/src/plan/validate.rs

```diff
@@ -46,5 +47,151 @@ pub fn validate(schema_map: &HashMap<String, Schema>, statement: Statement) -> R
         }
     }
 
-    Ok(statement)
+    Ok(())
+}
+
+pub enum ValidationContext<'a> {
+    Data {
+        table_name: &'a String,
+        alias: Option<&'a TableAlias>,
+        schema: &'a Schema,
+        next: Option<Rc<ValidationContext<'a>>>,
+    },
+    Bridge {
+        left: Rc<ValidationContext<'a>>,
+        right: Rc<ValidationContext<'a>>,
+    },
+}
+
+impl<'a> ValidationContext<'a> {
+    fn new(
+        table_name: &'a String,
+        alias: Option<&'a TableAlias>,
+        schema: &'a Schema,
+        next: Option<Rc<ValidationContext<'a>>>,
+    ) -> Self {
+        Self::Data {
+            table_name,
+            alias,
+            schema,
+            next,
+        }
+    }
+
+    fn concat(
+        left: Option<Rc<ValidationContext<'a>>>,
+        right: Option<Rc<ValidationContext<'a>>>,
+    ) -> Option<Rc<Self>> {
+        match (left, right) {
+            (Some(left), Some(right)) => Some(Rc::new(Self::Bridge { left, right })),
+            (context @ Some(_), None) | (None, context @ Some(_)) => context,
+            (None, None) => None,
+        }
+    }
+
+    fn count(&self, column_name: &String) -> i32 {
```

Comment:
```suggestion
    fn count(&self, column_name: &str) -> usize {
```

It doesn't seem to convert to a negative number.
Consider `usize`